[core][telemetry/01] migrate python telemetry recorder to opentelemetry

can-anyscale opened the pull request:
This is a series of PR to migrate metric collection from opencencus to openlemetry. For context, ray has two main parts that instrument and record metric through opencencus: c++ and python part. 

The python part consists of three components: 
- A singleton python object (let's call it the TelemetryMetricRecorder) that stores all telemetry that got sent to it in memories. 
- A PrometheusServer that holds this TelemetryMetricRecorder object. The server takes a snapshot of this object, convert telemetries into prometheus metrics and persist the metrics every few seconds.
- A DashboardAgent, which runs as a cronjob on every ray node, that collects node-level telemetries and send them to the TelemetryMetricRecorder

The c++ part consists of one component (to be deep dived more):
- A cronjob runs on every ray worker and send telemetries to TelemetryMetricRecorder via grpc messages.

----

This PR migrates the TelemetryMetricRecorder from using opencencus to opentelemetry. The migration is also control via a flag (which turns on the new stuff by default). Fairly small PR (most files are auto-generated).

Test:
- CI + new unit test
- test_metrics_agent.py + test_reporter.py has a lot of e2e tests and they still work
- e2e test in anyscale (go/cantel01)

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

aslonnie reviewed (approved):
approve for dependency change.

edoakes commented:
🙌

edoakes commented:
@MengjinYan can you review this pls?

can-anyscale commented on python/ray/_private/telemetry/open_telemetry_metric_recorder.py:
```diff
@@ -0,0 +1,80 @@
+                    for tag_set, val in observations
+                ]
+
+            instrument = self.meter.create_observable_gauge(
```
according to chatgpt you'are correct yes, for gauge, if there are observation with the same attributes it will just pick up the last one

MengjinYan replied:
So in this case, I guess my question is wondering is there are reason why we need a separate dictionary to cache the latest value, instead of directly recording the value in the corresponding instrument and let the default behavior to do the aggregation?

MengjinYan replied:
I guess another followup question I have is, if we foresee the `OpenTelemetryMetricRecorder` to be the entry point for all open telemetry metrics, I'm wondering how we plan to add the other types of instruments and aggregations?

can-anyscale replied:
> is there are reason why we need a separate dictionary to cache the latest value, instead of directly recording the value in the corresponding instrument 

my understanding is that the instrument object in opentelemetry is stateless; instead it offer a null pointer via a `callback` function that can point to anything; users then need manage the data themselves and point this `callback` function to their managed data; the callback function is invoked at collection time periodically

MengjinYan replied:
> my understanding is that the instrument object in opentelemetry is stateless; instead it offer a null pointer via a callback function that can point to anything; users then need manage the data themselves and point this callback function to their managed data; the callback function is invoked at collection time periodically

Sorry to be pedantic here, but From my understanding, there are 2 ways to record raw metrics to the open telemetry instrument: 
(1) manual instrument: directly calling the `Add` or `Record` method on the open instrument and the aggregation will be done by the default views; 
(2) observable instrument: having a callback to collect the metric whenever, the metric is scraped or collected

(1) is more of a push model where the raw metrics are pushed to the instrument and (2) is more of a pull model where the metrics will only be pulled when the collection happens. And I think in our case, the metrics are pushed to the metrics recorder so (1) might be better choice than (2)

MengjinYan replied:
Discussed offline. Here are the conclusion: 
* From what we see, open telemetry doesn't have a gauge type for manual instrument. The idea behind the design is that, for the gauge metrics, the values of the metrics should be obtained from the source at the time the metrics are being collected by the external reader (Prometheus in our case). And the mechanism should be achieved by the observable instrument mechanism.
* At the same time, the current implementation is done using manual instrument because open census doesn't support the observable instrument mechanism. 
* The migration can be implemented with 2 options:
  * Option 1: Directly redesign of the pipeline to use the observable instrument
  * Option 2: Do the implementation as of this PR to keep the manual instrument logic for now
* The benefit of redesign is mainly on the performance side: less metric cache in memory, less calls to gether metrics; At the same time, the downside is that because the pipeline deep integrated with not only core but other libraries (e.g. data), the whole migration design would take a lot of effort.
* Therefore, we will do option 2 in the first place and if get more data regarding the performance gains of option 1, we can decide on do a redesign in a later phase.

MengjinYan commented on python/ray/_private/telemetry/open_telemetry_metric_recorder.py:
```diff
@@ -0,0 +1,80 @@
+                        f"Failed to record metric {gauge.name} with value {value} with tags {tags!r} and global tags {global_tags!r} due to: {e!r}"
+                    )
+
+    def _record_gauge(self, gauge: Gauge, value: float, tags: dict):
```
Chatted offline. The Guage class here doesn't use any functions from the open census View class. Instead, it actually creates a View as a field in the class. And in the last step of the migration, when we are removing the open census dependency, we will remove the open census related fields here and make the class telemetry platform agnostic.

can-anyscale replied:
totally

can-anyscale commented:
@MengjinYan's comments

can-anyscale commented:
@MengjinYan: I changed `_observations_by_gauge_name` back to the version that deduplicates tags instead of simply appending data points. Previously, I reset` _observations_by_gauge_name` at collection time (triggered by the /metrics endpoint), but this broke `test_reporter` in a way that is updated recently by @dayshah: the test calls /metrics twice in a row and the second call returns no data (see [here](https://github.com/ray-project/ray/blob/master/python/ray/dashboard/modules/reporter/tests/test_reporter.py#L250)). To keep the behavior as close as it was, I removed the reset of _observations_by_gauge_name. It now needs to be a map rather than a vector of data points to prevent memory leaks.

We can discuss more in person next week perhaps if this makes no sense.

MengjinYan reviewed (approved):
Thanks!
